### PR TITLE
StringUtils concat and join

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Added variadic `concat` and `join` functions to `StringUtils`.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [X] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [X] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [X] This PR adds tests that were used to verify all changes:
  - [X] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13728/

**Comment by goedderz** [comment]
image

File: lib/Basics/StringUtils.h [hunk 1]
```diff
@@ -418,6 +418,101 @@ std::string decodeHex(std::string const& value);
 void escapeRegexParams(std::string& out, const char* ptr, size_t length);
 std::string escapeRegexParams(std::string const& in);
 
+namespace impl {
```
Comment (maierlars) [inline lib/Basics/StringUtils.h:421]: Normally we call this `detail` but it's fine for me.

File: lib/Basics/StringUtils.h [hunk 2]
```diff
@@ -418,6 +418,101 @@ std::string decodeHex(std::string const& value);
...
+}
+
+template <typename Iter, typename... Iters>
+auto joinImplIter(std::string_view delim, std::pair<Iter, Iter>&& head,
+                  std::pair<Iters, Iters>&&... tail) -> std::string {
+  auto result = std::string{};
+
+  auto const valueSizes = std::distance(head.first, head.second) +
+                          (std::distance(tail.first, tail.second) + ... + 0);
+  auto const delimSizes = sizeof...(Iters) * delim.size();
```
Comment (maierlars) [inline lib/Basics/StringUtils.h:469]: shouldn't there be `sizeof...(Iters) - 1` many delims?

File: tests/Basics/StringUtilsTest.cpp [hunk 3]
```diff
@@ -348,3 +348,34 @@ TEST_F(StringUtilsTest, test_encodeURLComponent) {
...
+  EXPECT_EQ("", StringUtils::joinT(", "sv, ""sv));
+  EXPECT_EQ("42", StringUtils::joinT(", "sv, 42));
+  EXPECT_EQ("2.500000", StringUtils::joinT(", "sv, 2.5));
+  EXPECT_EQ(", ", StringUtils::joinT(", "sv, "", ""));
+  EXPECT_EQ(", , ", StringUtils::joinT(", "sv, "", "", ""));
+  EXPECT_EQ("hello, world", StringUtils::joinT(", "sv, "hello", "world"));
+  EXPECT_EQ("hello, world", StringUtils::joinT(""sv, "hello", ", ", "world"));
+  EXPECT_EQ("cstr 42 stdstr view", StringUtils::joinT(""sv, "cstr ", 42, " stdstr "s, "view"sv));
+  EXPECT_EQ("cstr, 42, stdstr, view", StringUtils::joinT(", "sv, "cstr", 42, "stdstr"s, "view"sv));
+}
```
Comment (maierlars) [inline tests/Basics/StringUtilsTest.cpp:381]: maybe one could assert the capacity of the string to match the expected capacity.

**Review by maierlars — APPROVED** [review]

File: lib/Basics/StringUtils.h (hunk 2 above)
Comment (goedderz) [reply lib/Basics/StringUtils.h:469]: No, there are `1 + sizeof...(Iters)` many elements: note `head` and `tail`.

File: tests/Basics/StringUtilsTest.cpp (hunk 3 above)
Comment (goedderz) [reply tests/Basics/StringUtilsTest.cpp:381]: good idea.

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13735/

File: lib/Basics/StringUtils.h (hunk 2 above)
Comment (maierlars) [reply lib/Basics/StringUtils.h:472]: 👍🏽

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13737/

**Comment by jsteemann** [comment]
With this PR, is it possible to drop the previously existing implementations of `join` and `concat`?

**Comment by goedderz** [comment]
Depends on the usages. The new versions do not allow for a dynamic (at runtime) number of arguments, while the old ones do. If that's not needed, we can drop the old ones (and maybe drop the trailing `T` on the new ones). The call sites would have to be adapted, as they're not drop-in replacements. I can check the callers later this week, if there's any dynamic usage.

**Comment by goedderz** [comment]
I've immediately found several dynamic usages, so the existing implementations cannot be dropped.

**Comment by jsteemann** [comment]
Ok, if you could check that would be nice! Thanks.

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13752/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13755/

**Comment by goedderz** [comment]
image

**Pull request merged**